rootfs/apply.go

```diff
@@ -0,0 +1,136 @@
+package rootfs
+
+import (
+	"io"
+	"io/ioutil"
+
+	"github.com/docker/containerd"
+	"github.com/docker/containerd/log"
+	"github.com/docker/distribution"
+	"github.com/docker/docker/pkg/archive"
+	"github.com/opencontainers/go-digest"
+	"github.com/pkg/errors"
+)
+
+type Snapshotter interface {
+	Prepare(key, parent string) ([]containerd.Mount, error)
```

Comment:
@AkihiroSuda @samuelkarp Because of the lifecycle of a snapshot, having an active object here for the snapshot transaction doesn't make a whole lot of sense. Most of the data associated with `key` (the active transaction) will be disk bound. If we have these active object, there will be temptation to place inmemory metadata on the active object (see the `SnapshotManager` for this anti-pattern), which will lead to inconsistent metadata between drivers. If we force these all to be _just_ paths, we will keep the implementations simple and avoid bloat.

The main impetus for this is that a the lifecycle of a snapshot transaction will usually outlive the caller of these methods. If we have an object to represent this, there will be incorrect implementations that put cleanup into a Close method that doesn't necessarily belong there. The cleanup should be mostly isolated to `Rollback` and `Commit`, which may be called without calling `Prepare` in the same process.

Regarding godocs for these, please refer to the existing godocs for the snapshot manager/driver implementations. We'll get this all resolved as we come closer to a working end to end implementation.